commands/build.go

```diff
@@ -434,7 +437,245 @@ func buildCmd(dockerCli command.Cli, rootOpts *rootOptions) *cobra.Command {
 			options.contextPath = args[0]
 			options.builder = rootOpts.builder
 			cmd.Flags().VisitAll(checkWarnedFlags)
-			return runBuild(dockerCli, options)
+			if !isExperimental() || !options.detach {
+				return runBuild(dockerCli, options)
+			}
+
+			if options.debug {
+				logrus.SetLevel(logrus.DebugLevel)
+			}
+
+			var cInfo ControlInfo
+			cStr := os.Getenv(controlFDEnv)
+			if cStr != "" {
+				c, err := strconv.ParseInt(cStr, 10, 64)
+				if err != nil {
+					return err
+				}
+				file := os.NewFile(uintptr(c), "pipe")
+				defer file.Close()
+				if err := json.NewDecoder(file).Decode(&cInfo); err != nil {
+					return err
+				}
+				if cInfo.Mode == "" || cInfo.Root == "" {
+					return fmt.Errorf("received invalid control info: %+v", cInfo)
+				}
+			}
+			logrus.Debugf("control info: %+v", cInfo)
+
+			switch cInfo.Mode {
+			case "":
+				//
+				// step 0: buildx client
+				//
+				ctx := context.TODO()
+
+				// Process options
+				if options.quiet && options.progress != "auto" && options.progress != "quiet" {
+					return errors.Errorf("progress=%s and quiet cannot be used together", options.progress)
+				} else if options.quiet {
+					options.progress = "quiet"
+				}
+				if options.invoke != "" && (options.dockerfileName == "-" || options.contextPath == "-") {
+					// stdin must be usable for monitor
+					return errors.Errorf("Dockerfile or context from stdin is not supported with invoke")
+				}
+
+				// determine and perpare root dir for buildx server instance
+				rootDir := options.root
+				if rootDir == "" {
+					var err error
+					rootDir, err = rootDataDir()
+					if err != nil {
+						return err
+					}
+				}
+				if rootDir == "" {
+					return fmt.Errorf("buildx root dir must be determined")
+				}
+				if err := os.MkdirAll(rootDir, 0700); err != nil {
+					return err
+				}
+				instanceRoot, err := os.MkdirTemp(rootDir, "buildx")
```

Comment:
why is this called instanceRoot, doesn't look like it is instance specific?

Reply:
nvm, I think this is invocation instance, not builder instance. Bit confusing to use the same terms.